test/unit/src/metal/list/sort.cpp

```diff
@@ -3,47 +3,64 @@
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M), VALUE(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M), NUMBER(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M), PAIR(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M), LIST(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M), MAP(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, VALUE(M), LAMBDA(N)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), VALUE(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), NUMBER(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), PAIR(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), LIST(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), MAP(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), LAMBDA(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, NUMBER(M), LAMBDA(_)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), VALUE(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), NUMBER(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), PAIR(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), LIST(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), MAP(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), LAMBDA(N)>), (BOOL(N == 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, PAIR(M), LAMBDA(_)>), (FALSE)); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M)>), (BOOL(M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), VALUE(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), NUMBER(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), PAIR(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LIST(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), MAP(N)>), (FALSE)); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LAMBDA(N)>), (BOOL(N == 2 || M < 2))); \
     CHECK((metal::is_invocable<metal::lambda<metal::sort>, LIST(M), LAMBDA(_)>), (BOOL(M < 2))); \
+    CHECK((metal::is_invocable<metal::lambda<metal::sort>, MAP(M)>), (BOOL(M < 2))); \
```

Comment:
Caring about the tests made me think about the consequences of this change more thoroughly. In particular, this line for `M = 1` is nasty: It is probably not what the user intends, but it compiles.

I tried to apply the _easy-to-use,-hard-to-misuse_ rule here, but I do not see how we can achieve both (convenient default for the ordering relation, as expected from `std::sort` on the one hand, and exclusion of this nasty case on the other hand). For a moment I considered a solution that treats single-value lists specially (`invoke<lbd, x, x>` just to be sure), but that turns out to be a bad idea.

Do you share my concerns regarding the nasty case? What do you propose?

Reply:
> In particular, this line for `M = 1` is nasty

Very well observed, but this is not an issue that manifests itself only because we added default ordering to `metal::sort` - for an arbitrary `lbd` it is [still true](https://github.com/brunocodutra/metal/blob/master/test/unit/src/metal/list/sort.cpp#L31) that `is_invocable<lambda<sort>, map<pair<void, void>>, lambda<lbd>>::value == true`, even for a nasty `lbd` such as:

```cpp
template<typename... args>
using lbd = static_assert(sizeof...(args)  + 1 == 0, "always fails");
```

If we draw a parallel to runtime algorithms, specifying a function that unconditionally throws an exception as the predicate doesn't mean `std::sort` fails to sort a sequence of size 1.

Based on all that, I'd say this is fine.